packages/common-scripts/src/deploy.ts

```diff
@@ -0,0 +1,558 @@
+    config: options.config,
+  });
+
+  const output: Cell = {
+    cell_output: {
+      capacity: "0x0",
+      lock: options.outputScriptLock,
+    },
+    data: bytesToHex(options.scriptBinary),
+  };
+
+  txSkeleton = updateOutputs(txSkeleton, output);
+  txSkeleton = updateCellDeps(txSkeleton, options.config);
+  txSkeleton = await completeTx(txSkeleton, fromAddress, options.config);
+
+  const scriptConfig = getScriptConfig(txSkeleton, 0);
+
+  return {
+    txSkeleton,
+    scriptConfig,
+  };
+}
+
+/**
+ * Generate txSkeleton for writing binary data to CKB via Type ID, usually for deploying contracts.
+ * Deploying via Type ID makes it possible to upgrade contract, for more information about Type ID, please check: https://xuejie.space/2020_02_03_introduction_to_ckb_script_programming_type_id/
+ * This generator only supports `SECP256K1_BLAKE160` and `SECP256K1_BLAKE160_MULTISIG` currently.
+ *
+ * @param options
+ */
+export async function generateDeployWithTypeIdTx(
+  options: DeployOptions
+): Promise<TypeIDDeployResult> {
+  let txSkeleton = TransactionSkeleton({ cellProvider: options.cellProvider });
+
+  const fromAddress = generateAddress(options.outputScriptLock, {
+    config: options.config,
+  });
+
+  const [resolved] = await findCellsByLock(
+    options.outputScriptLock,
+    options.cellProvider
+  );
+  if (!resolved) throw new Error(`${fromAddress} has no live ckb`);
+
+  const typeId = utils.generateTypeIdScript(
+    { previous_output: resolved.out_point!, since: "0x0" },
+    "0x0"
+  );
+  const output: Cell = {
+    cell_output: {
+      capacity: "0x0",
+      lock: options.outputScriptLock,
+      type: typeId,
+    },
+    data: bytesToHex(options.scriptBinary),
+  };
+
+  txSkeleton = updateOutputs(txSkeleton, output);
+  txSkeleton = updateCellDeps(txSkeleton, options.config);
+  txSkeleton = await completeTx(txSkeleton, fromAddress, options.config);
+
+  const scriptConfig = getScriptConfig(txSkeleton, 0);
+
+  return {
+    txSkeleton,
+    scriptConfig,
+    typeId,
+  };
+}
+
+export async function generateUpgradeTypeIdDataTx(
+  options: UpgradeOptions
+): Promise<DeployResult> {
+  let txSkeleton = TransactionSkeleton({ cellProvider: options.cellProvider });
+
+  const fromAddress = generateAddress(options.outputScriptLock, {
+    config: options.config,
+  });
+
+  const collector = options.cellProvider.collector({ type: options.typeId });
+  const cells: Cell[] = [];
+  for await (const cell of collector.collect()) {
+    cells.push(cell);
+  }
+  if (cells.length !== 1) throw new Error("the typeid maybe wrong");
+
+  const deployedCell = cells[0];
+  txSkeleton = txSkeleton.update("inputs", (inputs) => {
+    return inputs.push(deployedCell);
+  });
+
+  const output: Cell = {
+    cell_output: {
+      capacity: "0x0",
+      lock: options.outputScriptLock,
+      type: options.typeId,
+    },
+    data: bytesToHex(options.scriptBinary),
+  };
+
+  txSkeleton = updateOutputs(txSkeleton, output);
```

Comment:
if the contract is reduced in size when updating, will the capacity be consumed by using `updateOutput` directly?